// mlflow-tracing/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SpanStatusCode = exports.MlflowClient = exports.withSpan = exports.trace = exports.startSpan = exports.init = exports.flushTraces = exports.updateCurrentTrace = exports.getCurrentActiveSpan = exports.getLastActiveTraceId = void 0;
const config_1 = require("./core/config");
Object.defineProperty(exports, "init", { enumerable: true, get: function () { return config_1.init; } });
const api_1 = require("./core/api");
Object.defineProperty(exports, "getLastActiveTraceId", { enumerable: true, get: function () { return api_1.getLastActiveTraceId; } });
Object.defineProperty(exports, "getCurrentActiveSpan", { enumerable: true, get: function () { return api_1.getCurrentActiveSpan; } });
Object.defineProperty(exports, "updateCurrentTrace", { enumerable: true, get: function () { return api_1.updateCurrentTrace; } });
Object.defineProperty(exports, "startSpan", { enumerable: true, get: function () { return api_1.startSpan; } });
Object.defineProperty(exports, "trace", { enumerable: true, get: function () { return api_1.trace; } });
Object.defineProperty(exports, "withSpan", { enumerable: true, get: function () { return api_1.withSpan; } });
const provider_1 = require("./core/provider");
Object.defineProperty(exports, "flushTraces", { enumerable: true, get: function () { return provider_1.flushTraces; } });
const clients_1 = require("./clients");
Object.defineProperty(exports, "MlflowClient", { enumerable: true, get: function () { return clients_1.MlflowClient; } });
// Export entities
__exportStar(require("./core/constants"), exports);
var span_status_1 = require("./core/entities/span_status");
Object.defineProperty(exports, "SpanStatusCode", { enumerable: true, get: function () { return span_status_1.SpanStatusCode; } });
