// mlflow-tracing/dist/core/entities/trace_location.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TraceLocationType = void 0;
exports.createTraceLocationFromExperimentId = createTraceLocationFromExperimentId;
/**
 * Types of trace locations
 */
var TraceLocationType;
(function (TraceLocationType) {
    /**
     * Unspecified trace location type
     */
    TraceLocationType["TRACE_LOCATION_TYPE_UNSPECIFIED"] = "TRACE_LOCATION_TYPE_UNSPECIFIED";
    /**
     * Trace is stored in an MLflow experiment
     */
    TraceLocationType["MLFLOW_EXPERIMENT"] = "MLFLOW_EXPERIMENT";
    /**
     * Trace is stored in a Databricks inference table
     */
    TraceLocationType["INFERENCE_TABLE"] = "INFERENCE_TABLE";
})(TraceLocationType || (exports.TraceLocationType = TraceLocationType = {}));
/**
 * Create a TraceLocation from an experiment ID
 * @param experimentId The ID of the MLflow experiment
 */
function createTraceLocationFromExperimentId(experimentId) {
    return {
        type: TraceLocationType.MLFLOW_EXPERIMENT,
        mlflowExperiment: {
            experimentId: experimentId
        }
    };
}
