// mlflow-tracing/dist/core/entities/trace_state.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TraceState = void 0;
exports.fromOtelStatus = fromOtelStatus;
const api_1 = require("@opentelemetry/api");
/**
 * Enum representing the state of a trace
 */
var TraceState;
(function (TraceState) {
    /**
     * Unspecified trace state
     */
    TraceState["STATE_UNSPECIFIED"] = "STATE_UNSPECIFIED";
    /**
     * Trace successfully completed
     */
    TraceState["OK"] = "OK";
    /**
     * Trace encountered an error
     */
    TraceState["ERROR"] = "ERROR";
    /**
     * Trace is currently in progress
     */
    TraceState["IN_PROGRESS"] = "IN_PROGRESS";
})(TraceState || (exports.TraceState = TraceState = {}));
/**
 * Convert OpenTelemetry status code to MLflow TraceState
 * @param statusCode OpenTelemetry status code
 */
function fromOtelStatus(statusCode) {
    switch (statusCode) {
        case api_1.SpanStatusCode.OK:
            return TraceState.OK;
        case api_1.SpanStatusCode.ERROR:
            return TraceState.ERROR;
        default:
            return TraceState.STATE_UNSPECIFIED;
    }
}
