// mlflow-tracing/dist/clients/spec.js
"use strict";
/**
 * MLflow API Request/Response Specifications
 *
 * This module defines the types and interfaces for MLflow API communication,
 * including request payloads and response structures for all trace-related endpoints.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetCredentialsForTraceDataDownload = exports.GetCredentialsForTraceDataUpload = exports.DeleteExperiment = exports.CreateExperiment = exports.GetTraceInfoV3 = exports.StartTraceV3 = void 0;
/**
 * Create a new TraceInfo entity in the backend.
 */
var StartTraceV3;
(function (StartTraceV3) {
    StartTraceV3.getEndpoint = (host) => `${host}/api/3.0/mlflow/traces`;
})(StartTraceV3 || (exports.StartTraceV3 = StartTraceV3 = {}));
/**
 * Get the TraceInfo entity for a given trace ID.
 */
var GetTraceInfoV3;
(function (GetTraceInfoV3) {
    GetTraceInfoV3.getEndpoint = (host, traceId) => `${host}/api/3.0/mlflow/traces/${traceId}`;
})(GetTraceInfoV3 || (exports.GetTraceInfoV3 = GetTraceInfoV3 = {}));
/** Create Experiment (used for testing) */
var CreateExperiment;
(function (CreateExperiment) {
    CreateExperiment.getEndpoint = (host) => `${host}/api/2.0/mlflow/experiments/create`;
})(CreateExperiment || (exports.CreateExperiment = CreateExperiment = {}));
/** Delete Experiment (used for testing) */
var DeleteExperiment;
(function (DeleteExperiment) {
    DeleteExperiment.getEndpoint = (host) => `${host}/api/2.0/mlflow/experiments/delete`;
})(DeleteExperiment || (exports.DeleteExperiment = DeleteExperiment = {}));
/**
 * Get credentials for uploading trace data to the artifact store. Only used for Databricks.
 */
var GetCredentialsForTraceDataUpload;
(function (GetCredentialsForTraceDataUpload) {
    GetCredentialsForTraceDataUpload.getEndpoint = (host, traceId) => `${host}/api/2.0/mlflow/traces/${traceId}/credentials-for-data-upload`;
})(GetCredentialsForTraceDataUpload || (exports.GetCredentialsForTraceDataUpload = GetCredentialsForTraceDataUpload = {}));
/**
 * Get credentials for downloading trace data from the artifact store. Only used for Databricks.
 */
var GetCredentialsForTraceDataDownload;
(function (GetCredentialsForTraceDataDownload) {
    GetCredentialsForTraceDataDownload.getEndpoint = (host, traceId) => `${host}/api/2.0/mlflow/traces/${traceId}/credentials-for-data-download`;
})(GetCredentialsForTraceDataDownload || (exports.GetCredentialsForTraceDataDownload = GetCredentialsForTraceDataDownload = {}));
